### PR TITLE
Fixed compilation on macOS Sierra

### PR DESCRIPTION
**Pull request by eddyc**
Added flags to binding.gyp for Clang to find the correct C++ library.

### CODE REVIEW

**Comment by joeroback** [comment]
Hrmm, adding '-stdlib=libc++' is mac specific and errors on linux

```
> atom-clang@1.0.13 install /home/jroback/Github/atom-clang
> node-gyp rebuild

make: Entering directory '/home/jroback/Github/atom-clang/build'
  CXX(target) Release/obj.target/clang/src/clang_translationunit.o
clang.target.mk:122: recipe for target 'Release/obj.target/clang/src/clang_translationunit.o' failed
make: Leaving directory '/home/jroback/Github/atom-clang/build'

g++: error: unrecognized command line option ‘-stdlib=libc++’
make: *** [Release/obj.target/clang/src/clang_translationunit.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/share/atom/resources/app/apm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 4.4.0-53-generic
gyp ERR! command "/usr/share/atom/resources/app/apm/bin/node" "/usr/share/atom/resources/app/apm/node_modules/.bin/node-gyp" "rebuild"
gyp ERR! cwd /home/jroback/Github/atom-clang
gyp ERR! node -v v4.4.5
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 

npm ERR! Linux 4.4.0-53-generic
npm ERR! argv "/usr/share/atom/resources/app/apm/bin/node" "/usr/share/atom/resources/app/apm/node_modules/npm/bin/npm-cli.js" "--globalconfig" "/home/jroback/.atom/.apm/.apmrc" "--userconfig" "/home/jroback/.atom/.apmrc" "install" "--runtime=electron" "--target=1.3.13" "--arch=x64"
npm ERR! node v4.4.5
npm ERR! npm  v3.10.5
npm ERR! code ELIFECYCLE
npm ERR! atom-clang@1.0.13 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the atom-clang@1.0.13 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the atom-clang package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs atom-clang
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls atom-clang
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/jroback/Github/atom-clang/npm-debug.log
```

**Comment by eddyc** [comment]
Ah OK, there seems to be conditions available for the binding.gyp file to configure properly for different OSs, I'll have a look

**Comment by joeroback** [comment]
In addition, using clang-3.9 on macOS it also results in error, as that seems to be an option specific to only Apple's LLVM/CLANG...

```
> atom-clang@1.0.13 install /Users/jroback/Dropbox/Projects/Github/atom-clang
> node-gyp rebuild

  CXX(target) Release/obj.target/clang/src/clang_translationunit.o

c++: error: unrecognized command line option '-stdlib=libc++'
make: *** [Release/obj.target/clang/src/clang_translationunit.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Applications/Atom.app/Contents/Resources/app/apm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 16.1.0
gyp ERR! command "/Applications/Atom.app/Contents/Resources/app/apm/bin/node" "/Applications/Atom.app/Contents/Resources/app/apm/node_modules/.bin/node-gyp" "rebuild"
gyp ERR! cwd /Users/jroback/Dropbox/Projects/Github/atom-clang
gyp ERR! node -v v4.4.5
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 

npm ERR! Darwin 16.1.0
npm ERR! argv "/Applications/Atom.app/Contents/Resources/app/apm/bin/node" "/Applications/Atom.app/Contents/Resources/app/apm/node_modules/npm/bin/npm-cli.js" "--globalconfig" "/Users/jroback/.atom/.apm/.apmrc" "--userconfig" "/Users/jroback/.atom/.apmrc" "install" "--runtime=electron" "--target=1.3.13" "--arch=x64"
npm ERR! node v4.4.5
npm ERR! npm  v3.10.5
npm ERR! code ELIFECYCLE
npm ERR! atom-clang@1.0.13 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the atom-clang@1.0.13 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the atom-clang package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs atom-clang
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls atom-clang
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/jroback/Dropbox/Projects/Github/atom-clang/npm-debug.log
```

**Comment by joeroback** [comment]
I'll have to think about how to handle this.

**Comment by eddyc** [comment]
Fair enough, I added the conditions there and it still works building on my Mac so maybe that would be worth using in some form.

File: binding.gyp [hunk 1]
```diff
@@ -28,11 +28,22 @@
...
+        ['OS=="mac"', {
+           "cflags_cc": [
+           "-g",
+           "-std=c++11",
+           "-stdlib=libc++",
+           "-O2"
+          ],
+        }],
+        ['OS=="linux"', {
+           "cflags_cc": [
```
Comment (joeroback) [inline binding.gyp]: these flags are not used on mac, that is why there is an entirely separate "xcode_settings", because apple is dumb

**Comment by joeroback** [comment]
yea this totally helps figure out your mac compile issue,  I just have to make sure not to break the other platforms/compilers :)

**Comment by joeroback** [comment]
damn this is where binding.gyp really shows how much it sucks as a reinvention of the wheel.. why doesn't npm just not use CMake? binding.gyp is not smart enough to know that its using the Apple compiler, which needs `-stdlib=libc++` for C++11, but if you are using G++ and CLANG (not Apple's), they both do not. CMake can handle this situation.

If Apple's clang included the libclang headers, we wouldn't even need to install clang, but it doesn't, it only includes libclang.dylib, not the headers required to build.

So I am inclined at this point to not do anything, as most people using this package are using GCC or CLANG (not Apple's compiler), until binding.gyp gets smarter and is able to tell which compiler it is using. Otherwise, if I add this flag, Apple's compiler will work, but GCC/CLANG will fail. Catch 22.

Argh!

**Comment by joeroback** [comment]
fixed in https://github.com/joeroback/atom-clang/pull/18

**Pull request closed without merging**